File: src/main/java/com/indra87g/rewards/DailyRewardManager.java
package com.indra87g.rewards;

import cn.nukkit.Player;
import cn.nukkit.Server;
import cn.nukkit.item.Item;
import cn.nukkit.potion.Effect;
import cn.nukkit.utils.Config;
import cn.nukkit.utils.TextFormat;
import me.onebone.economyapi.EconomyAPI;

import java.io.File;
import java.time.*;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DailyRewardManager {
    private final Config dataConfig;
    private final Config mainConfig;
    private final Map<String, String> messages = new HashMap<>();

    private final ZoneId zoneId;
    private final LocalTime resetTime;

    public DailyRewardManager(File configFile, File dataFolder) {
        this.mainConfig = new Config(configFile, Config.YAML);

        File dataFile = new File(dataFolder, "players.yml");
        this.dataConfig = new Config(dataFile, Config.YAML);

        this.zoneId = ZoneId.of(mainConfig.getString("timezone", "UTC"));
        this.resetTime = LocalTime.parse(mainConfig.getString("claim_reset", "00:00"));

        if (mainConfig.get("messages") instanceof Map) {
            for (Map.Entry<String, Object> entry : ((Map<String, Object>) mainConfig.get("messages")).entrySet()) {
                if (entry.getValue() instanceof String) {
                    messages.put(entry.getKey(), (String) entry.getValue());
                }
            }
        }
    }

    public boolean claimReward(Player player) {
        String name = player.getName().toLowerCase();
        long lastClaimSeconds = dataConfig.getLong(name + ".last_claim", 0);
        int streak = dataConfig.getInt(name + ".streak", 0);

        ZonedDateTime now = ZonedDateTime.now(zoneId);
        ZonedDateTime lastReset = getLastReset(now);

        if (lastClaimSeconds >= lastReset.toEpochSecond()) {
            ZonedDateTime nextReset = lastReset.plusDays(1);
            long secondsRemaining = now.until(nextReset, ChronoUnit.SECONDS);
            long hours = secondsRemaining / 3600;
            long minutes = (secondsRemaining % 3600) / 60;
            String timeRemaining = String.format("%dh %dm", hours, minutes);
            sendMessage(player, "already_claimed", "{time}", timeRemaining);
            return false;
        }

        ZonedDateTime previousReset = lastReset.minusDays(1);
        if (lastClaimSeconds >= previousReset.toEpochSecond()) {
            streak++;
        } else {
            streak = 1;
        }

        List<List<Map<String, Object>>> rewardsList = (List<List<Map<String, Object>>>) mainConfig.getList("rewards");
        if (rewardsList == null || rewardsList.isEmpty()) {
            sendMessage(player, "not_available");
            return false;
        }

        dataConfig.set(name + ".last_claim", now.toEpochSecond());
        dataConfig.set(name + ".streak", streak);
        dataConfig.save();

        int rewardIndex = (streak - 1) % rewardsList.size();
        List<Map<String, Object>> rewards = rewardsList.get(rewardIndex);

        giveRewards(player, rewards);
        sendMessage(player, "success", "{day}", String.valueOf(streak));

        if (streak == rewardsList.size()) {
            sendMessage(player, "rewards_claimed");
        }

        return true;
    }

    private void giveRewards(Player player, List<Map<String, Object>> rewards) {
        if (rewards == null) return;
        for (Map<String, Object> reward : rewards) {
            String type = reward.get("type").toString().toLowerCase();
            try {
                switch (type) {
                    case "item":
                        int id = ((Number) reward.get("id")).intValue();
                        int meta = reward.containsKey("meta") ? ((Number) reward.get("meta")).intValue() : 0;
                        int amount = ((Number) reward.get("amount")).intValue();
                        player.getInventory().addItem(Item.get(id, meta, amount));
                        break;
                    case "money":
                        double money = ((Number) reward.get("amount")).doubleValue();
                        EconomyAPI.getInstance().addMoney(player, money);
                        break;
                    case "effect":
                        Effect effect = Effect.getEffectByName(reward.get("effect").toString().toUpperCase());
                        if (effect != null) {
                            int duration = ((Number) reward.get("duration")).intValue();
                            int amplifier = ((Number) reward.get("amplifier")).intValue();
                            effect.setDuration(duration * 20);
                            effect.setAmplifier(amplifier - 1);
                            player.addEffect(effect);
                        }
                        break;
                    case "exp":
                        int exp = ((Number) reward.get("amount")).intValue();
                        player.addExperience(exp);
                        break;
                    case "command":
                        String cmd = reward.get("command").toString().replace("%player%", player.getName());
                        Server.getInstance().dispatchCommand(Server.getInstance().getConsoleSender(), cmd);
                        break;
                }
            } catch (Exception e) {
                player.getServer().getLogger().error("Failed to give daily reward of type '" + type + "' to " + player.getName(), e);
                player.sendMessage(TextFormat.colorize("&cAn error occurred while giving you a reward. Please contact an admin."));
            }
        }
    }

    // ... existing methods ...
    public void sendMessage(Player player, String key, String... replacements) {
        String message = messages.getOrDefault(key, "&cMessage not found in daily_rewards.yml: " + key);
        if (replacements.length % 2 == 0) {
            for (int i = 0; i < replacements.length; i += 2) {
                message = message.replace(replacements[i], replacements[i + 1]);
            }
        }
        player.sendMessage(TextFormat.colorize(message));
    }

    public ZonedDateTime getLastReset(ZonedDateTime now) {
        ZonedDateTime todayReset = now.with(resetTime);
        return now.isBefore(todayReset) ? todayReset.minusDays(1) : todayReset;
    }

    public int getStreak(Player player) {
        return dataConfig.getInt(player.getName().toLowerCase() + ".streak", 0);
    }

    public ZonedDateTime getNextResetTime() {
        return getLastReset(ZonedDateTime.now(zoneId)).plusDays(1);
    }

    public long getLastClaimTime(Player player) {
        return dataConfig.getLong(player.getName().toLowerCase() + ".last_claim", 0);
    }
}

File: src/main/kotlin/com/indra87g/rewards/DailyRewardManager.kt
package com.indra87g.rewards

import cn.nukkit.Player
import cn.nukkit.Server
import cn.nukkit.item.Item
import cn.nukkit.potion.Effect
import cn.nukkit.utils.Config
import com.indra87g.utils.MessageHandler
import me.onebone.economyapi.EconomyAPI
import java.io.File
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

class DailyRewardManager(configFile: File, dataFolder: File) {
    private val mainConfig: Config = Config(configFile, Config.YAML)
    private val dataConfig: Config
    private val zoneId: ZoneId
    private val resetTime: LocalTime

    val now: ZonedDateTime
        get() = ZonedDateTime.now(zoneId)

    init {
        val dataFile = File(dataFolder, "players.yml")
        this.dataConfig = Config(dataFile, Config.YAML)
        this.zoneId = ZoneId.of(mainConfig.getString("timezone", "UTC"))
        this.resetTime = LocalTime.parse(mainConfig.getString("claim_reset", "00:00"))
    }

    fun claimReward(player: Player): Boolean {
        val name = player.name.lowercase()
        val lastClaimSeconds = dataConfig.getLong("$name.last_claim", 0)
        var streak = dataConfig.getInt("$name.streak", 0)
        val now = ZonedDateTime.now(zoneId)
        val lastReset = getResetTime(now)
        val prevReset = lastReset.minusDays(1)

        if (lastClaimSeconds >= lastReset.toEpochSecond()) {
            val nextReset = lastReset.plusDays(1)
            val secondsRemaining = now.until(nextReset, ChronoUnit.SECONDS)
            val hours = secondsRemaining / 3600
            val minutes = secondsRemaining % 3600 / 60
            val timeRemaining = String.format("%dh %dm", hours, minutes)
            MessageHandler.sendMessage(player, "daily_reward_already_claimed", "{time}", timeRemaining)
            return false
        }

        streak = if (lastClaimSeconds >= prevReset.toEpochSecond() && lastClaimSeconds < lastReset.toEpochSecond()) {
            streak + 1
        } else {
            1
        }

        dataConfig.set("$name.last_claim", now.toEpochSecond())
        dataConfig.set("$name.streak", streak)
        dataConfig.save()

        var rewardGiven = false
        val todayDate = now.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
        if (mainConfig.exists("special_rewards.$todayDate")) {
            giveRewards(player, mainConfig.getMapList("special_rewards.$todayDate"))
            MessageHandler.sendMessage(player, "daily_reward_special", "{date}", todayDate)
            rewardGiven = true
        }

        val dayOfWeek = now.dayOfWeek.value // Monday=1, Sunday=7
        val rewardKey = "rewards.reward$dayOfWeek"
        if (mainConfig.isList(rewardKey)) {
            giveRewards(player, mainConfig.getMapList(rewardKey))
            val dayName = now.dayOfWeek.toString()
            MessageHandler.sendMessage(player, "daily_reward_success", "{day}", dayName)
            rewardGiven = true
        }

        if (mainConfig.exists("streak_rewards.$streak")) {
            giveRewards(player, mainConfig.getMapList("streak_rewards.$streak"))
            MessageHandler.sendMessage(player, "daily_reward_streak_bonus", "{streak}", streak.toString())
            rewardGiven = true
        }

        if (!rewardGiven) {
            MessageHandler.sendMessage(player, "daily_reward_not_available")
        }
        return rewardGiven
    }

    private fun giveRewards(player: Player, rewards: List<Map<*, *>>) {
        rewards.forEach { reward ->
            when (reward["type"].toString().lowercase()) {
                "item" -> {
                    val id = reward["id"].toString().toInt()
                    val amount = reward["amount"].toString().toInt()
                    player.inventory.addItem(Item(id, 0, amount))
                }
                "money" -> {
                    val money = reward["amount"].toString().toDouble()
                    EconomyAPI.getInstance().addMoney(player, money)
                }
                "effect" -> {
                    val effect = Effect.getEffectByName(reward["effect"].toString().uppercase())
                    if (effect != null) {
                        effect.duration = reward["duration"].toString().toInt() * 20
                        effect.setAmplifier(reward["amplifier"].toString().toInt())
                        player.addEffect(effect)
                    }
                }
                "exp" -> {
                    val exp = reward["amount"].toString().toInt()
                    player.addExperience(exp)
                }
                "command" -> {
                    val cmd = reward["command"].toString().replace("%player%", player.name)
                    Server.getInstance().dispatchCommand(Server.getInstance().consoleSender, cmd)
                }
            }
        }
    }

    private fun getResetTime(now: ZonedDateTime): ZonedDateTime {
        return now.withHour(resetTime.hour).withMinute(resetTime.minute).withSecond(0).withNano(0)
    }

    fun getStreak(player: Player): Int {
        return dataConfig.getInt("${player.name.lowercase()}.streak", 0)
    }

    fun hasSpecialReward(date: String): Boolean {
        return mainConfig.exists("special_rewards.$date")
    }
}
